### PR TITLE
Revisions

### PR DESCRIPTION
**Pull request by wasade**
These changes actually put in support for contexts. I didn't realize these commits hadn't been pushed in. This allows for housing different sets of processed data, supporting a 1-many relationship between the sample metadata and the sample data.

The tests have also been decomposed. An important note on the decisions surrounding the test environment. The click commands at this time are not wrapped in unit tests. This was a pragmatic decision, as it was easier and faster to write integration tests from the command line (i.e., the contents of `test.sh` and `test_failures.sh`). This will begin to shift as the unit tests will be easier to manage, however it is a lower priority target at this time.

A critical note: this PR does not include support for the prep ID. Working on that in a moment.

cc @josenavas @antgonza

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,45 +3,92 @@
...
 
-Load BIOM tables and sample metadata from a lot of studies into Redis to facilitate search using observation IDs or metadata.
+Load BIOM tables and sample metadata from a lot of studies into Redis to facilitate search using observation IDs or metadata. redbiom is intended as a caching layer. The command line application, for many of the commands, supports Unix pipes allowing for a fair bit of composability. As a hypothetical example, let's say we had loaded a bunch of microbiome data picked againts Greengenes 13_8 at 97% (and into an appropriately named context). To obtain a BIOM table representing all samples in which either OTUs 367523 or 3064251 exist, you can perform the following set of querys:
 
-It is safe to say the interfaces are not stable.
+    $ redbiom search observations --context "greengenes-13_8-97" 367523 3064251 | redbiom fetch samples --from - --output result.biom
+
+Additionally, assuming a comprehensive set of metadata are available, you can obtain summarized environment information:
+    
+    $ redbiom search observations --context "greengenes-13_8-97" 367523 3064251 | redbiom summarize samples --from - --category "env_biome"
```
Comment (antgonza) [inline README.md]: do you think that most users will get the magic dash?

File: README.md [hunk 2]
```diff
@@ -3,45 +3,92 @@
...
 
 # Installation
 
-Needs Redis (tested with 3.2.6) and Webdis (just clone the repo). Also depends on BIOM (tested on 2.1.5-dev), Pandas (tested on 0.19.0), Click (tested on 6.6) and scipy (whatever BIOM is happy with).
+Needs Redis (tested with 3.2.6) and Webdis (just clone the repo). Also depends on BIOM (tested on 2.1.5), Pandas (tested on 0.19.0), Click (required >= 6.7) and scipy (whatever BIOM is happy with). It is not necessary to have super user access.
```
Comment (antgonza) [inline README.md]: What about adding links to the required repos/versions?

File: README.md [hunk 3]
```diff
@@ -3,45 +3,92 @@
...
-	$ ./redbiom admin load-sample-metadata --metadata path/to/qiime/compat/mapping.txt
+	$ redbiom admin load-sample-metadata --metadata path/to/qiime/compat/mapping.txt
 
-Next, we'll load up associations between every single observation in the BIOM table to all the samples its found in. This will create Redis sets and can be accessed using keys of the form `samples:<observation_id>`. 
+**IMPORTANT**: values containing `/` characters cannot be represented the forward slash is used to denote arguments with Webdis. At present, these values are omitted. This is more generally a problem for dates which have not been normalized into an ISO standard. See issue #9.
 
-	$ ./redbiom admin load-observations --table /path/to/biom/table.biom
+**IMPORTANT**: values which appear to be null are not stored. The set of values currently considered nulls are: 
+    
+    {'Not applicable', 'Unknown', 'Unspecified', 'Missing: Not collected',
```
Comment (antgonza) [inline README.md]: Should _all_ these values actually be null? Users _might_ want to look for values that were not provided due to restricted access?

File: README.md [hunk 4]
```diff
@@ -3,45 +3,92 @@
...
+The test harness is broken into multiple components, and are driven by a `Makefile`. This was done initially to be pragmatic as it was easier to write integration tests than unit tests for the `click` commands. That, and I didn't have wifi at the time in order to refer to the docs. These tests can be found in `test.sh` which is composed of "positive" tests and `test_failures.sh` which is composed of "negative" tests. The difference being that the positive tests will fail if any command results in a nonzero exit status, whereas the negative tests expect a nonzero exit status (and really, the decision was to avoid unsetting "-e"). Additional tests which validate some of the Redis contents can be found in `redbiom/tests/test_rest.py`. These are neither unit tests nor integration tests, but simply exercise the behind-the-scenes REST interface. Last, there are some unit tests placed under `redbiom/tests/test_util.py`. Tests have additionally not been written _yet_ for the `requests` calls also due to a lack of access to documentation at the time of writing, however due to the integration tests, there is a high confidence the methods are operating correctly sans some edge cases.
```
Comment (antgonza) [inline README.md]: lack of `requests` tests should be an issue too ...

File: README.md (hunk 1 above)
Comment (wasade) [inline README.md]: That's POSIX-compliant. That's literally how you specify the stdin file descriptor. There isn't another means that I'm aware of to support reading from stdin within click unless operating outside of click, which is doable, but that is not a good solution.

File: README.md (hunk 3 above)
Comment (wasade) [inline README.md]: Can reduce, what do you suggest we retain?

File: README.md (hunk 4 above)
Comment (wasade) [inline README.md]: Those tests are already part of this PR, I just didn't have the chance to revise the readme again.

File: README.md (hunk 1 above)
Comment (antgonza) [inline README.md]: k

File: README.md (hunk 3 above)
Comment (antgonza) [inline README.md]: I _think_ is fine to just remove/null: 'null', 'NULL', 'no_data', 'None', 'nan' ...

File: README.md (hunk 4 above)
Comment (antgonza) [inline README.md]: K, can you update the readme?

File: README.md (hunk 2 above)
Comment (wasade) [inline README.md]: done

File: README.md (hunk 1 above)
Comment (wasade) [inline README.md]: There is a boilerplate replacement that can be done which I'll add. It's somewhat outside of click so it's could bite us later but perhaps it wont

File: README.md (hunk 3 above)
Comment (wasade) [inline README.md]: does qiita have a defined set of null values?

Comment (josenavas) [inline README.md]: We no longer infer null values in Qiita. In the ideal world, users should be using the EBI standard...

Comment (wasade) [inline README.md]: okay

Comment (wasade) [inline README.md]: ...i guess on that note, should this PR just be merged as is? It doesn't sound like there is a consensus on what is to be considered null. The present list covers values which do not represent useful or public information, unless there is a clear use case for finding samples which are denoted as, say, restricted or not provided or something?

**Comment by wasade** [comment]
Thanks!
